docs/l1/fundamentals/snap_sync.md

```diff
@@ -0,0 +1,321 @@
+
+![Fast Sync Retaking Example - 1](snap_sync/Fast%20Sync%20Retaking%20Example%20-%201.png)
+
+In the example, even if we find that node { hash: 0x317f, path: 0 } is correct, we still need to check all its children in the DB (in this case none are present).
+
+To solve the second problem, we introduce an optimization, which is called the "Membatch". This allows us to maintain a new invariant:
+
+> If a node is present in the DB, then that node and all its children must be present.
+
+This removes the need to explore entire subtrees: when walking down the trie, if a node is in the DB, the whole subtree can be skipped.
+
+To maintain this invariant, we do the following:
+
+- When we get a new node, we don't immediately store it in the database. We keep track of the amount of every node's children that are not yet in the database. As long as it's not zero, we keep it in a separate in-memory structure "Membatch" instead of on the db.
+- When a node has all of its children in the db, we commit it and recursively go up the tree to see if its parent needs to be commited, etc.
+
+Example of a possible state after stoping fast sync due to staleness with membatch.
+
+![Fast Sync Membatch Example - 1](snap_sync/Fast%20Sync%20Membatch%20Example%20-%201.png)
+
+### Speeding up: Snap-Sync
+
+Fast-sync is slow (ethrex fast-sync of a fresh “hoodi” state takes ~45 minutes—4.5× slower than snap-sync). To accelerate it, we use a key property of fast-sync:
+> Fast-sync can “heal” any partial trie that obeys the invariant, even if that trie is not consistent with any real on-chain state.
+
+Snap-sync exploits this:
+
+- download only the leaves (the accounts and storage slots) from any recent state,
+- assemble a trie from these leaves,
+- run fast-sync (“healing”) to repair it into a consistent trie.
+
+For our code, we call the fast-sync the healing step.
+
+Example run:
+
+\- We download the 4 accounts in the state trie from two different blocks
+
+![Snap Sync Leaves - 1](snap_sync/Snap%20Sync%20Leaves%20-%201.png)
+
+\- We rebuild the trie
+
+![Snap Sync Healing - 1](snap_sync/Snap%20Sync%20Healing%20-%201.png)
+
+\- We run the healing algorithm and get a correct tree at the end of it
+
+This method alone provides up to a 4-5 times boost in performance, as computing the trie is way faster than downloading it.
+
+## Implementation
+
+Generalized Flowchart of snapsync
+
+![Flow - Snap Sync](snap_sync/Flow%20-%20Snap%20Sync.png)
+
+### Flags
+
+When testing snap sync there are flags to take into account:
+
+- If the `SKIP_START_SNAP_SYNC` environment variable is set and isn't empty, it will skip the step of downloading the leaves and will immediatly begin healing. This simulates the behaviour of fast-sync.
+
+- If debug assertions are on, the program will validate that the entire state and storage tries are valid by traversing the entire trie and recomputing the roots. If any is found to be wrong, it will print an error and exit the program.
```

Comment:
> - If debug assertions are on, the program will validate that the entire state and storage tries are valid by traversing the entire trie and recomputing the roots. If any is found to be wrong, it will print an error and exit the program.

I'd add here something like "this is also used for testing purposes; a validation error here means there is a 100% a bug in snap sync."